Diff commit on branch

stoivo opened the pull request:
I created `f + m` in branch interface. This displays a graph of commits that on one of the branches but not  on the active. It have all the same functions as graph view have. Not sure if that is good or bad.

it looks like this: 
![screen shot 2015-06-03 at 22 57 43](https://cloud.githubusercontent.com/assets/3492040/7971300/0f315aa8-0a44-11e5-91b1-cb24947b74bd.png)


This is not done completed but I would like to get you feedback.

divmain commented on docs/branch_mgmt.md:
```diff
@@ -82,10 +82,15 @@ The selected branch will be merged into the active branch.  If the merge cannot
+
+#### Diff commits selected branch against active (`f + m`)
+
+A scratch view will be opened, showing the commit diff between the selected branch and the active branch.
```
```
A scratch view will be opened, showing all commits present on the selected branch but not the active branch, and vice versa.
```

divmain commented on Default.sublime-keymap:
```diff
@@ -471,13 +471,20 @@
     },
     {
-        "keys": ["f"],
+        "keys": ["f", "d"],
```
No need to have two-key bindings here.  Change this to `"keys": ["H"]`.

stoivo replied:
Fine, is there any logic for that? History?
My question is why 'H'?

divmain replied:
Yeah, I was thinking "diff history" as the short-description.  Generally, unless grouping very-related commands together (like stash commands), I try to stick to single key-presses.  This new command relates to the diff command in that it sort of does something similar.  Sticking to single key-bindings (when possible) also avoids the issue of mistakenly invoking the wrong command when the first key wasn't registered.

stoivo replied:
Make sense.

divmain commented on docs/branch_mgmt.md:
```diff
@@ -82,10 +82,15 @@ The selected branch will be merged into the active branch.  If the merge cannot
 A scratch view will be opened, showing the diff between the selected branch and the active branch.
 
+
+#### Diff commits selected branch against active (`f + m`)
```
```
#### Diff history of selected and active branches (`H`)
```

divmain commented on core/interfaces/branch.py:
```diff
@@ -612,3 +612,60 @@ class GsBranchesFetchCommand(TextCommand, GitCommand):
         self.view.window().run_command("gs_fetch")
+
+
+class GsBranchesDiffCommitsBranch(TextCommand, GitCommand):
```
`GsBranchesDiffBranchHistory`

divmain commented on core/interfaces/branch.py:
```diff
@@ -612,3 +612,60 @@ class GsBranchesFetchCommand(TextCommand, GitCommand):
+        view.settings().set("word_wrap", False)
+        view.set_name("BRANCH COMMIT COMPARISON")
+        view.set_scratch(True)
+        view.run_command("gs_replace_view_text", {"text": diff_contents, "nuke_cursors": True})
```
This can be cleaned up with `util.get_read_only_view`.

stoivo replied:
maybe it would be good to be able to pass a hash of option that we can use to set settings like 
``` pyhon3
view.settings().set("git_savvy.repo_path", repo_path)
view.settings().set("word_wrap", False)
```

divmain replied:
Most of these read-only views behave in similar ways.  The `repo_path` will always be passed, I can't think of a time when you'd want `word_wrap` to be `True`, etc.  So unless there's a reason, I'd prefer for the helper method to set sane defaults and only make exceptions when necessary.

stoivo replied:
We do want to use `word_wrap` on committing or looking at a commit. Like from the graph.

divmain commented on core/interfaces/branch.py:
```diff
@@ -612,3 +612,60 @@ class GsBranchesFetchCommand(TextCommand, GitCommand):
+        args.pop()
+        args.append("{}..{}".format(branch1, branch2))
+        diff_contents += self.git(*args)
+        return diff_contents
```
Since this is something new, it may be best to add this functionality as a new interface, rather than the old-style view instantiation / view logic.  The branch-dashboard command would only call the interface constructor with the selected branch name.

divmain commented on core/interfaces/branch.py:
```diff
@@ -612,3 +612,60 @@ class GsBranchesFetchCommand(TextCommand, GitCommand):
+        args.append("{}..{}".format(branch2, branch1))
+        diff_contents += self.git(*args)
+        diff_contents += "\n\nCommits on {} and not on {}\n".format(branch2, branch1)
+        args.pop()
```
This is a little too clever.  Should be abstracted into a method that takes two args, `branchA`, and `branchB`, and returns the contents from that combination.

divmain commented:
This is a worthy feature!  Let me know once you've addressed my comments.  I may adapt what you've written here into an interface, which is where most of these things have been heading.

divmain commented:
Also, once you've made the requested changes, could you squash and rebase on top of latest master?  Looks like there is at least one conflict.

stoivo commented:
All I need to do is to Convert it to a interface. Is there anything more I should change or add to it?

divmain commented:
This looks good.  However, I'd like you to remove the `GsMoveCorsurToRemoteBranch` command and associated keybinding.  That is definitely a different feature, so should go in a separate PR.  And I have other plans for how navigation between sections should work, so this likely conflicts.

stoivo commented:
@divmain, now I removed GsMoveCorsurToRemoteBranch and rebased once more :)
Should I change it to a interface?

I whould like to have GsMoveCorsurToRemoteBranch somehow. Can I get  to know your plan?

divmain commented:
Thanks for this.  Merging now.  Can you ping me on Gitter to discuss `GsMoveCorsurToRemoteBranch`?  I need a refresher on what it did so that I can properly reply to your question.
